### PR TITLE
Textfiled support cursor v1.2

### PR DESCRIPTION
**Pull request by perminovVS**
Add StringUtils::StringUTF8 for easy manipulate utf8 string.
If TextField active(focused input text) show text cursor( default '|',
change TextFieldTTF::setCursorChar ).
Show/Hide interval CURSOR_TIME_SHOW_HIDE
Change position text cursor control key: KEY_HOME, KEY_END,
KEY_LEFT_ARROW, KEY_RIGHT_ARROW
CC_PLATFORM_MAC, CC_PLATFORM_WIN32, CC_PLATFORM_LINUX default use
cursor( For change state TextFieldTTF::setCursorUse )
Desktop implement dispatchControlKey
WinRT support key repeat and implement dispatchControlKey
Text cursor worked only on LabelType::TTF

new version it: #13903 , for easy merge

Example use: http://youtu.be/g6jpmaip1fY

### CODE REVIEW

**Comment by zilongshanren** [comment]
@perminovVS  :+1:  Sweet!

**Comment by zilongshanren** [comment]
@perminovVS 
It seems the cursor support is only available for desktop?

File: cocos/base/ccUTF8.h [hunk 1]
```diff
@@ -142,6 +142,43 @@ CC_DLL unsigned int getIndexOfLastNotChar16(const std::vector<char16_t>& str, ch
...
+    };
+    typedef std::vector<CharUTF8> CharUTF8Store;
+
+    StringUTF8();
+    StringUTF8(const std::string& newStr);
+    explicit StringUTF8(const StringUTF8& copyStrUtf8);
+    ~StringUTF8();
+
+    std::size_t length() const;
+    void set(const std::string& newStr);
```
Comment (zilongshanren) [inline cocos/base/ccUTF8.h]: The `set` API seems not intuitive.. What's the intent of this API?

File: cocos/base/ccUTF8.h [hunk 2]
```diff
@@ -142,6 +142,43 @@ CC_DLL unsigned int getIndexOfLastNotChar16(const std::vector<char16_t>& str, ch
...
+    struct CharUTF8
+    {
+        std::string _char;
+        bool isAnsi() { return _char.size() == 1; }
+    };
+    typedef std::vector<CharUTF8> CharUTF8Store;
+
+    StringUTF8();
+    StringUTF8(const std::string& newStr);
+    explicit StringUTF8(const StringUTF8& copyStrUtf8);
```
Comment (zilongshanren) [inline cocos/base/ccUTF8.h]: if we define a user defined copy constructor, we might need to use the "Rule of Three" or "Rule of Five"...

File: cocos/base/ccUTF8.h [hunk 3]
```diff
@@ -142,6 +142,43 @@ CC_DLL unsigned int getIndexOfLastNotChar16(const std::vector<char16_t>& str, ch
...
 CC_DLL std::vector<char16_t> getChar16VectorFromUTF16String(const std::u16string& utf16);
 
+
+
+/**
+* Utf8 sequence
+* Store all utf8 chars as std::string
+* Build from std::string
+*/
+class CC_DLL StringUTF8
```
Comment (zilongshanren) [inline cocos/base/ccUTF8.h:152]: One more question: Do we really need the StringUTF8 encapsulation? I think we have already provided many utf8, utf16 string conversion functions, why bother to provide another StringUTF8 class?
Thanks.

**Comment by perminovVS** [comment]
@zilongshanren CC_PLATFORM_MAC, CC_PLATFORM_WIN32, CC_PLATFORM_LINUX default use
cursor, other platform need change state TextFieldTTF::setCursorUse(true)

File: cocos/base/ccUTF8.h (hunk 1 above)
Comment (perminovVS) [inline cocos/base/ccUTF8.h]: @zilongshanren set new value from std::string

Comment (zilongshanren) [inline cocos/base/ccUTF8.h]: @perminovVS  It's `append` or `replace`?

**Comment by zilongshanren** [comment]
@perminovVS 
Thanks for the clarification, I will do more test to this PR.

File: cocos/base/ccUTF8.h (hunk 3 above)
Comment (perminovVS) [inline cocos/base/ccUTF8.h:152]: StringUTF8 for easy manipulate sequence utf8 chars;
example from TextFieldTTF

```
StringUtils::StringUTF8 stringUTF8;

stringUTF8.set(_inputText);
stringUTF8.insert(_cursorPosition, insert);
```

File: cocos/base/ccUTF8.h (hunk 1 above)
Comment (perminovVS) [inline cocos/base/ccUTF8.h]: @zilongshanren set new values, as replace all

File: cocos/2d/CCTextFieldTTF.h [hunk 4]
```diff
@@ -230,6 +236,32 @@ class CC_DLL TextFieldTTF : public Label, public IMEDelegate
 
     virtual void visit(Renderer *renderer, const Mat4 &parentTransform, uint32_t parentFlags) override;
 
+    virtual void update(float delta) override;
+
+    /**
+    * Set enable cursor use.
+    * @js NA
+    */
+    void setCursorUse(bool value);
```
Comment (zilongshanren) [inline cocos/2d/CCTextFieldTTF.h]: `setCursorEnabled` might be better than `setCursorUse`.

File: cocos/2d/CCTextFieldTTF.h [hunk 5]
```diff
@@ -252,6 +287,21 @@ class CC_DLL TextFieldTTF : public Label, public IMEDelegate
 
     bool _secureTextEntry;
 
+    // Need use cursor
+    bool _cursorUse;
```
Comment (zilongshanren) [inline cocos/2d/CCTextFieldTTF.h]: change `_cursorUse` to `_cursorEnabled`.

File: cocos/base/ccUTF8.h (hunk 1 above)
Comment (zilongshanren) [inline cocos/base/ccUTF8.h]: so @perminovVS  how about about renaming it to `replace` or sth  else?

**Comment by zilongshanren** [comment]
@perminovVS 
Now I can't click to change cursor position, only arrow key is working. Thoughts?

**Comment by perminovVS** [comment]
@zilongshanren yes, need manualy call setCursorFromPoint.
Need listen mouse or touch event and call it.

In UITextField can add this functional.

**Comment by perminovVS** [comment]
@zilongshanren review now

**Comment by zilongshanren** [comment]
@perminovVS  I'm OK with this PR now.  Thanks.

**Comment by 0xJimmyBeh** [comment]
@perminovVS  @zilongshanren 

May i know how to use this cursor feature on UITextField?

I tried to implement with Cocos2d-JS v3.8.1 and generated the bindings.
But i can't see the cursor after `setCursorEnabled(true);` in iOS.

Thank you.

**Comment by siarsky** [comment]
@Zinitter 
I've looked into the cocos2d_tests project - the class TextInputTests is a good starting point with one very crucial error regarding a new cursor functionality.

The TextInputTests contains a method onEnter() - which calls
`KeyboardNotificationLayer::onEnter();`
but does not call:
`pTextField->onEnter()`
in which the schedule "update(float delta)" method will be resumed (default = paused). If onEnter() is not called on TextFieldTTF, the TextFieldTTF::update(float delta) method will not be called and the cursor will not blink.

Maybe you just do not see the cursor because it is a non-blinkin pipe sign (|) at the end of the string due to not calling TextFieldTTF::onEnter()...

After calling onEnter the cursor functionality works for me in iOs and Android as well.

**Pull request merged**